proposals/0000-hybrid-output.md

```diff
@@ -0,0 +1,139 @@
+- Start Date: 2022-10-25
+- Reference Issues: TODO
+- Implementation PR: <!-- leave empty -->
+
+# Summary
+
+Currently, Astro's [`output`](https://docs.astro.build/en/reference/configuration-reference/#output) enables developers to switch between `'static'` (default) and `'server'` outputs. One of our most requested features is to enable route-level control of `output` so that some routes can be server-rendered while others are statically generated.
+
+# Example
+
+This RFC proposes a new, opt-in `output` mode named `'hybrid'`.
+
+```js
+import { defineConfig } from 'astro/config'
+import node from '@astrojs/node'
+
+export default defineConfig({
+  output: 'hybrid',
+  adapter: node()
+})
+```
+
+For individual routes, `'static'` will be the default. To opt-in to `'server'` output, users may `export const output: 'static' | 'server'`.
+
+```astro
+---
+export const output = 'server'
+
+const text = Astro.url.searchParams.get('text') ?? 'Hello world!';
+---
+
+<h1 set:html={text} />
+```
+
+# Motivation
+
+Since Astro v1.0.0 was released, we've gotten consistent feedback that more granular control over `output` is a requirement for scaling projects. Users would like to statically render some routes and server render others, reducing the need for full static rebuilds for frequently changing data.
+
+One very common use-case is a fully static site that also exposes `api/` endpoints which are handled on the server. Another common example is a largely static site that can generate a few dynamic pages at request time (for example, a user profile page).
+
+The eventual solution should enable granular control over `output` on a per-route basis. Since Astro defaults to `'static'` output, this should remain the default in the eventual solution as well.
+
+# Detailed design
+
+## Public API
+
+There are two public API surfaces to this RFC:
+
+- Updating the `astro.config.mjs` to accept a new `output` value, `'hybrid'`. The existing values of `'static' | 'server'` would remain supported. A value of `'hybrid'` will enabled detection for the following feature:
+- Allowing routes (files in `pages/`) to self-declare a granular `output` value. The existing `output` values of `'static' | 'server'` would both be valid.
+
+**This declaration must be statically analyzable!**
+Only string literal values will be accepted. This is because a route cannot dynamically be static or server rendered based on an incoming request—this value _must_ be known at build time. Other declaration values will throw an error and fail the build. This mirrors similar constraints that Vite has for compile-time (macro) features like dynamic `import()` and `import.meta.glob`.
+
+```astro
+---
+// Valid – constant string literal
+export const output = 'server'
+// Invalid – cannot be determined at compile-time
+export const output = static ? 'static' : 'server';
+export const output = value;
+export const output = 'ser' + 'ver';
+export let output = undefined; output = 'server';
```

Comment:
I would assume this is not allowed either: `export let output = 'server'`. Allowing let/var means someone might try to mutate it. Since that's harder to restrict I think it makes sense to require the user of `const`.

Reply:
That seems fair!